### PR TITLE
Make better use of macros.

### PR DESCRIPTION
**Pull request by jsha**
Add checked_ptr_ref! and checked_ptr_mut! to eliminate the
repeated null-pointer checking code we had at the beginning of each
function.

Revise the ffi_panic_boundary_* macros to be implemented in terms of
ffi_panic_boundary_generic!, which takes a return value as an argument.

Remove some `const` modifiers on parameters that are actually treated
as mutable (rustls_client_session_read, rustls_client_session_write,
etc).

I think overall this improves clarity by removing boilerplate. For instance
the wants_read and is_handshaking methods are much more clearly
plain passthroughs now. But feel free to tell me "you're getting carried
away with macros!"

### CODE REVIEW

File: src/lib.rs
Comment (alex) [inline src/lib.rs:114]: Is there a reason not to implement this in terms of the other variant? So: `checked_ptr_ref!($var, $typ, rustls_result::NullParameter)`?

Comment (jsha) [reply src/lib.rs]: Good call, done.

File: src/lib.rs
Comment (tgeoghegan) [inline src/lib.rs:113]: Does this need to be `as *mut $typ` or would `as *$typ` suffice? Put otherwise: since we're ultimately getting an immutable reference, why do we need a mutable reference to `var`?

File: src/lib.rs
Comment (tgeoghegan) [inline src/lib.rs:185]: 👍🏻 I like that the `unsafe` section is made smaller.

File: src/lib.rs
Comment (tgeoghegan) [inline src/lib.rs:197]: nit: for consistency with other usages of `checked_ptr_ref` and `checked_ptr_mut`, and what I think is your preferred style, you might do `let config: &mut ClientConfig = checked_ptr_mut!(config, ClientConfig);`. On the other hand, that would repeat the type `ClientConfig` twice, so perhaps this usage is better.

File: src/lib.rs
Comment (tgeoghegan) [inline src/lib.rs:250]: Looking at this usage, I don't love that `checked_ptr_ref!` can `return` an error value and bail out of the function, because I don't think the control flow is obvious when I read this function. Since there's no `?` or other error handling or branching, my assumption if I were scanning this code, perhaps seeking the line that produced the error message I found in a log, is that `checked_ptr_ref!` can't fail. What if instead of either evaluating to a `T` value or returning some value, `checked_ptr_ref!` evaluated to `Result<T>`, and then you'd do:
`let config = checked_ptr_ref!(config, ClientConfig, ())?;`
Which I think makes it more obvious that there's a potential failure case on this line. If you end up having to reintroduce the `match` statement at each callsite, then I think that would defeat the point of macro-ing away the checked conversion, but if it's just an extra `?`, I think it's worth it for clarity.

File: src/lib.rs
Comment (tgeoghegan) [inline src/lib.rs:353]:
```suggestion
        let session: &ClientSession = checked_ptr_ref!(session, ClientSession, false);
```

File: src/lib.rs
Comment (tgeoghegan) [inline src/lib.rs:107]: Galaxy brain review comment: could you write a single macro that matches on the literal `mut` in front of the `$typ:ty`?

```
macro_rules! checked_ptr {
    ( $var:ident, mut $typ:ty, $retval:expr ) => {
        // handle mutable case with .as_mut()
    };
    ( $var:ident, $typ:ty, $retval:expr ) => {
       // handle immutable case with .as_ref()
    }
}
```

Or maybe there's some other way to write arms that match on either `mut T` vs `T`? Then you wouldn't have to define two nearly identical macros for mut vs. ref. This is definitely code golf, and I'm more curious about whether it's possible than advocating for it as a good idea.

**Review by tgeoghegan — APPROVED** [review]
Approving because there's a couple typos that I'm sure you can fix without me doing another review, and then some ideas that you should feel free to ignore, so nothing that should block merging. This is neat, I'm enjoying learning more about Rust macros!

File: src/lib.rs
Comment (jsha) [reply src/lib.rs:185]: FWIW, it's not really. The `checked_ptr_mut!` implementation is one big unsafe blocks. So now we just have two smaller unsafe blocks rather than one big one.

File: src/lib.rs
Comment (jsha) [reply src/lib.rs:250]: I would love to do this. `?` only works in functions that return Result or Option. So it wouldn't work in this function that returns unit.

It seems like it *might* be possible for the functions that return rustls_result to return `Option<rustls_result>`, and get Rust to represent `None` as rustls_result::Ok (not sure how I would go about this but would do some research).

Another tweak that would provide much smaller benefit but much cheaper: I could rename these to start with `try_`, to remind people of the `try` macro that predated `?`: https://doc.rust-lang.org/std/macro.try.html. But given that's deprecated, it will probably resonate with fewer people over time.

File: src/lib.rs
Comment (jsha) [reply src/lib.rs:113]: Good catch. This actually shouldbe `*const $typ`.

File: src/lib.rs
Comment (jsha) [reply src/lib.rs:107]: This was a cool idea and it worked! I didn't wind up de-duplicating the macro definitions (probably could but I felt like it would be too fancy). But by adding `&` and `&mut` to the matches, the call sites for the macro more clearly express the type being cast to.

**Pull request merged**
